New extension: Granola AI Meeting Notes

RobErskine opened the pull request:
## Description
 
Saw the [AI Extensions Beta launch video](https://www.youtube.com/watch?v=sHIlFKKaq0A) and wanted to streamline Thomas' Granola example even further. This PR adds in new capabilities for interacting with Granola through a new extension.

- Added `New Note` command, which starts a new note and recording immediately in Granola
- Added `Search Notes` command, which allows you to view your notes in a list, view their details, copy their links, or copy their contents as HTML or Markdown
- Added `AI Notes` AI command, which allows you to use Raycast AI on top of Granola and other AI extensions

## Screencast
https://github.com/user-attachments/assets/4d79bd51-7b74-4985-87b8-18df9e116cff

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
![CleanShot 2025-03-06 at 16 04 17@2x](https://github.com/user-attachments/assets/9a1c1734-fed2-4ea6-96d8-d9919d6374b6)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used in the `README` are located outside the metadata folder if they were not generated with our [metadata tool](https://developers.raycast.com/basics/prepare-an-extension-for-store#how-to-use-it)

RobErskine commented:
dang @greptile-apps, thanks for the feedback. Going to move this over to Draft as I address the feedback.

RobErskine reviewed (commented):
Thanks for the help @greptile-apps. Please review the latest adjustments and let me know if there is anything else.

RobErskine reviewed (commented):
Mind giving me another once over, @Greptile-apps?

RobErskine reviewed (commented):
sorry i'm new to the greptile game and not sure how to get a new review to trigger.. thanks for your patience here. I think I have reviewed all of the feedback and resolved all the comments. Is there anything else that needs to happen to move this PR forward?

RobErskine commented on extensions/granola/src/utils/getAccessToken.ts:
```diff
@@ -0,0 +1,35 @@
+  const homeDirectory = os.homedir();
+
+  // Construct the path to the supabase.json file
+  const filePath = path.join(homeDirectory, "Library", "Application Support", "Granola", "supabase.json");
```
Granola is currently only available on Mac. It is not applicable to Windows/Linux users.

pernielsentikaer commented on extensions/granola/src/search-notes.tsx:
```diff
@@ -0,0 +1,136 @@
+            content={convertHtmlToMarkdown(notes)}
+            title="Copy Notes as Markdown"
+          />
+          <Action.CopyToClipboard icon={Icon.CodeBlock} content={notes} title="Copy Notes as Html" />
```
```suggestion
          <Action.CopyToClipboard icon={Icon.CodeBlock} content={notes} title="Copy Notes as HTML" />
```

RobErskine replied:
I had this as `HTML` but the linter was unhappy with it: `Prefer Title Case naming convention for action titles (e.g Copy to Clipboard).eslint[@raycast/prefer-title-case](https://github.com/raycast/eslint-plugin/blob/main/docs/rules/prefer-title-case.md)`. Running `npm run fix-lint --fix` reverts it back to `Html`. Is adding `{/* eslint-disable-next-line */}` acceptable here?

pernielsentikaer commented on extensions/granola/src/search-notes.tsx:
```diff
@@ -0,0 +1,136 @@
+            actions={
+              <ActionPanel>
+                <Action.Push
+                  title="Show Details"
```
Missing an icon

RobErskine commented:
Thanks for the feedback @pernielsentikaer! I've added in the icon the `Action.Push` as well as updated the copy to use `... as HTML` instead of `... as Html` using `{/* eslint-disable-next-line */}` so it doesn't get reverted automatically by the linter. If you have anything else please let me know.

RobErskine commented:
heya @pernielsentikaer, no rush but just wanted to check on the status of this new extension's PR and if there was anything I could do to help. I totally understand that there is an influx of submissions now, but I'm not totally sure of the process with the new automated Greptile and other security reviews. If there is anything I can do to move this forward please let me know.

RobErskine reviewed (commented):
Still getting used to Greptile.. just realizing I never viewed the final file in the PR with resolved comments. Going to submit another review here now.

pernielsentikaer commented:
Sorry for the longer response time for AI Extension; we're still sorting some things out internally 😊

There should be progress soon 😊

andreaselia commented:
@greptile-ai all good to go here now?

andreaselia commented:
@greptileai all good to go here?

andreaselia commented on extensions/granola/src/utils/fetchData.ts:
```diff
@@ -0,0 +1,39 @@
+      : undefined,
+  });
+
+  if (error) {
+    throw error;
+  }
```
Pretty sure this is caught when it's used in the other areas.

andreaselia commented on extensions/granola/src/utils/fetchData.ts:
```diff
@@ -0,0 +1,39 @@
+
+  const url = `https://api.granola.ai/v2/${route}`;
+
+  const { isLoading, data, revalidate } = useFetch<GetDocumentsResponse>(url, {
+    headers: accessToken
+      ? {
+          Authorization: `Bearer ${accessToken}`,
+        }
+      : undefined,
+  });
```
I mean, technically there's another property for this (`enabled` isn't what I'd use), but `undefined` should also avoid a not so nice response.

RobErskine commented:
Spoke to @andreaselia over DMs on Slack. Just dropped a commit based on some of his feedback:
- Implemented more resilient error handling when the access token does not exist
- Show the original user's notes, if they exist
- Feed the original user's notes to the AI Tool, if they exist
As always, if there is any feedback, please let me know!

RobErskine commented:
Connected with @andreaselia on one other small bit of feedback.
- Updated the rendering logic to include any of your original/handwritten notes in the Granola interface, as well as the AI generated notes.
- Updated the logic in the AI tool to check for this content before returning data.

Please check it out and lemme know if there is any other feedback 🚀

RobErskine commented:
Connected with @andreaselia one more time! Turns out My AI Tool implementation needed a small tweak to include original notes, as well as AI generated notes.

andreaselia reviewed (approved):
LGTM, thanks for working with me on the AI tool functionality @RobErskine - it is greatly appreciated!
